pkg/webhookadmission/webhook.go

```diff
@@ -0,0 +1,210 @@
+package webhookadmission
+
+import (
+	"context"
+	"crypto/tls"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/libopenstorage/stork/drivers/volume"
+	"github.com/portworx/sched-ops/k8s"
+	log "github.com/sirupsen/logrus"
+	"k8s.io/api/admission/v1beta1"
+	appv1 "k8s.io/api/apps/v1"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/tools/record"
+)
+
+const (
+	mutateWebHook   = "/mutate"
+	validateWebHook = "/validate"
+	certFile        = "/etc/webhook/certs/cert.pem"
+	certKeyFile     = "/etc/webhook/certs/key.pem"
+	schedulerPath   = "/spec/template/spec/schedulerName"
+	storkScheduler  = "stork"
+)
+
+// Controller for admission mutating webhook to initialise resources using px volume
+// with stork as scheduler
+type Controller struct {
+	Recorder record.EventRecorder
+	Driver   volume.Driver
+	server   *http.Server
+	lock     sync.Mutex
+	started  bool
+}
+
+// Serve method for webhook server
+func (c *Controller) serveHTTP(w http.ResponseWriter, req *http.Request) {
+	if strings.Contains(req.URL.Path, mutateWebHook) {
+		c.processMutateRequest(w, req)
+	} else {
+		http.Error(w, "Unsupported request", http.StatusNotFound)
+	}
+}
+
+// TODO: add logs to event recorder
+func (c *Controller) processMutateRequest(w http.ResponseWriter, req *http.Request) {
+	var admissionResponse *v1beta1.AdmissionResponse
+	admissionReview := v1beta1.AdmissionReview{}
+	arReq := admissionReview.Request
+	isPXResource := false
+
+	decoder := json.NewDecoder(req.Body)
+	defer func() {
+		if err := req.Body.Close(); err != nil {
+			log.Warnf("Error closing decoder")
+		}
+	}()
+	if err := decoder.Decode(&admissionReview); err != nil {
+		log.Errorf("Error decoding admission review request: %v", err)
+		c.Recorder.Event(&admissionReview, v1.EventTypeWarning, "Invalid admission review request", err.Error())
+		http.Error(w, "Decode error", http.StatusBadRequest)
+		return
+	}
+
+	// TODO: This log does not get reflected in stork pods , check other logger
+	log.Warnf("request received %v :%v", admissionReview.Kind, admissionReview)
+	switch arReq.Kind.Kind {
+	case "StatefulSet":
+		var ss appv1.StatefulSet
+		if err := json.Unmarshal(arReq.Object.Raw, &ss); err != nil {
+			log.Errorf("Could not unmarshal admission review object: %v", err)
+			c.Recorder.Event(&admissionReview, v1.EventTypeWarning, "could not unmarshal ar object", err.Error())
+		}
+		// check whether pod spec use px volume claims
+		volumes := ss.Spec.Template.Spec.Volumes
+		for _, v := range volumes {
+			pvc, err := k8s.Instance().GetPersistentVolumeClaim(v.PersistentVolumeClaim.ClaimName, ss.Namespace)
+			if err != nil {
+				c.Recorder.Event(&admissionReview, v1.EventTypeWarning, "Could not get pvc details %v", err.Error())
+				isPXResource = false
+				break
+			}
+			if c.Driver.OwnsPVC(pvc) {
+				isPXResource = true
+				break
+			}
+		}
+	case "Deployment":
```

Comment:
Also need this for pods since operators might not use deployments or statefulsets